2_process/src/observed_data_helpers_munge.R

```diff
@@ -0,0 +1,51 @@
+
+filter_observed_data <- function(obs_fn) {
+  
+  observed_dat <- read_feather(obs_fn)
+  
+  # Filter so that there is more than 1 temperature value 
+  # per date (more than one depth)
+  dates_to_keep <- observed_dat %>% 
+    group_by(date) %>%
+    summarize(count = n()) %>% 
+    filter(count > 1) %>% 
+    pull(date)
+  
+  observed_dat %>% 
+    filter(date %in% dates_to_keep) %>% 
+    # Discovered during `reformat_observed_data` that some sites did
+    # not have unique combinations of date & depth which causes the
+    # pivoted data to have lists instead of single values. To fix, I
+    # am averaging the temperature values if there are more than 1.
+    group_by(site_id, date, depth) %>% 
+    summarize(temp = mean(temp)) %>% 
+    ungroup()
+  
+}
+
+reformat_observed_data <- function(obs_data) {
+  
+  obs_data %>% 
+    # This is the date column name that `calculate_toha_per_lake` expects
+    rename(DateTime = date) %>% 
+    mutate(col_name = paste0("temp_", depth)) %>%  
+    arrange(depth) %>% 
+    # Reformat to be `temp_[depth]` as columns with one row per date
+    tidyr::pivot_wider(id_cols = c("site_id", "DateTime"), 
+                       names_from = col_name, 
+                       values_from = temp) 
+  
+}
+
+join_observed_data <- function(target_name, obs_data, irr_data_fn, k0_data_fn) {
+  
+  # Reformat to be `temp_[depth]` as columns with one row per date
+  obs_data %>% 
+    left_join(read_csv(irr_data_fn, col_types = cols()), by = c("DateTime" = "date"), progress = FALSE) %>%
+    rename(io = rad_0) %>% 
+    left_join(read_csv(k0_data_fn, col_types = cols()), by = c("DateTime" = "date"), progress = FALSE) %>% 
+    select(site_id, DateTime, io, kd, everything()) %>% 
+    write_feather(target_name)
+  
+}
```

Comment:
These three functions don't look like they would take a long time...do they? And/or is there another reason to treat them as separate steps with separate pipeline targets? If not, you could save some task table complexity and I/O time by writing a wrapper function for a single task-table step that calls each of these functions internally.

Reply:
I don't think they would take a long time. I was just trying to keep steps separated as much as possible. Maybe I can keep these 3 fxns separately but call them in one "process_data" type function? They mostly pass objects around anyways, so that would be easy.

Reply:
see https://github.com/USGS-R/lake-temperature-out/pull/27/commits/c2ed273c53e1fb92b868a1aca030d9fbad545731

Reply:
Yeah, I think that's a good approach - 3 functions wrapped with 1 wrapper function feels fairly lightweight to me (and still conceptually useful) relative to 3 steps per task.